fmn/consumer/consumer.py

```diff
@@ -49,16 +49,33 @@ def __init__(self, *args, **kwargs):
         log.debug("Loading rules from fmn.rules")
         self.valid_paths = fmn.lib.load_rules(root="fmn.rules")
 
-        self.cached_preferences = None
+        # Initialize our in-memory cache of the FMN preferences database
         self.cached_preferences_lock = threading.Lock()
+        self.cached_preferences = None
+        session = self.make_session()
+        self.refresh_cache(session)
+        session.close()
 
         log.debug("FMNConsumer initialized")
 
-    def refresh_cache(self, session, topic=None, msg=None):
-        log.info("Loading and caching preferences")
+    def refresh_cache(self, session, openid=None):
+        log.debug("Acquiring cached_preferences_lock")
         with self.cached_preferences_lock:
-            self.cached_preferences = fmn.lib.load_preferences(
-                session, self.hub.config, self.valid_paths, cull_disabled=True)
+            if not openid or self.cached_preferences is None:
+                log.info("Loading and caching all preferences for all users")
+                self.cached_preferences = fmn.lib.load_preferences(
+                    session, self.hub.config, self.valid_paths,
+                    cull_disabled=True)
+            else:
+                log.info("Loading and caching preferences for %r" % openid)
+                old_preferences = [p for p in self.cached_preferences
+                                   if p['user']['openid'] == openid]
+                new_preferences = fmn.lib.load_preferences(
+                    session, self.hub.config, self.valid_paths,
+                    cull_disabled=True, openid=openid)
+                self.cached_preferences.extend(new_preferences)
```

Comment:
Should we add the new pref after removing the old ones?

I am afraid if we get a conflict of identifiers we would remove the new pref we just added.